javalib/src/main/scala/java/util/ArrayList.scala

```diff
@@ -14,24 +14,51 @@ package java.util
 
 import java.lang.Cloneable
 import java.lang.Utils._
+import java.util.ScalaOps._
 
 import scala.scalajs._
+import scala.scalajs.LinkingInfo.targetPureWasm
 
-class ArrayList[E] private (private[ArrayList] val inner: js.Array[E])
+// _size keeps the track of the effective size of the underlying scala.Array for Wasm
+// in JS, we can use innerJS.length for it's size.
+class ArrayList[E] private (private[ArrayList] var inner: AnyRef, private var _size: Int)
     extends AbstractList[E] with RandomAccess with Cloneable with Serializable {
   self =>
 
+  private val innerJS =
+    if (targetPureWasm) null
+    else inner.asInstanceOf[js.Array[E]]
+
+  // To assign null to elements either when E <: AnyRef or AnyVal,
+  // we use an Array[Any] as an underlying storage and cast to E when needed.
+  private var innerWasm =
+    if (!targetPureWasm) null
+    else inner.asInstanceOf[Array[Any]]
+
   def this(initialCapacity: Int) = {
-    this(new js.Array[E])
+    this(
+      if (targetPureWasm)
+        new scala.Array[Any](Math.max(initialCapacity, 0))
```

Comment:
```suggestion
        new Array[Any](Math.max(initialCapacity, 0))
```
No need to add the `scala.` prefix. Same elsewhere.